src/System.Threading.Tasks.Extensions/tests/ManualResetValueTaskSource.cs

```diff
@@ -0,0 +1,171 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Runtime.ExceptionServices;
+using System.Threading.Tasks.Sources;
+
+namespace System.Threading.Tasks.Tests
+{
+    internal static class ManualResetValueTaskSource
+    {
+        public static ManualResetValueTaskSource<T> Completed<T>(T result, Exception error = null)
+        {
+            var vts = new ManualResetValueTaskSource<T>();
+            if (error != null)
+            {
+                vts.SetException(error);
+            }
+            else
+            {
+                vts.SetResult(result);
+            }
+            return vts;
+        }
+
+        public static ManualResetValueTaskSource<T> Delay<T>(int delayMs, T result, Exception error = null)
+        {
+            var vts = new ManualResetValueTaskSource<T>();
+            Task.Delay(delayMs).ContinueWith(_ =>
+            {
+                if (error != null)
+                {
+                    vts.SetException(error);
+                }
+                else
+                {
+                    vts.SetResult(result);
+                }
+            });
+            return vts;
+        }
+    }
+
+    internal sealed class ManualResetValueTaskSource<T> : IValueTaskSource<T>, IValueTaskSource
+    {
+        private static readonly Action<object> s_sentinel = new Action<object>(s => { });
+        private Action<object> _continuation;
+        private object _continuationState;
+        private SynchronizationContext _capturedContext;
+        private ExecutionContext _executionContext;
+        private bool _completed;
+        private T _result;
+        private ExceptionDispatchInfo _error;
+
+        public ValueTaskSourceStatus GetStatus(short token) =>
+            !_completed ? ValueTaskSourceStatus.Pending :
+            _error == null ? ValueTaskSourceStatus.Succeeded :
+            _error.SourceException is OperationCanceledException ? ValueTaskSourceStatus.Canceled :
+            ValueTaskSourceStatus.Faulted;
+
+        public T GetResult(short token)
+        {
+            if (!_completed)
+            {
+                throw new Exception("Not completed");
+            }
+
+            ExecutionContext ctx = _executionContext;
+            if (ctx != null)
+            {
+                ctx.Dispose();
+                _executionContext = null;
+            }
+
+            _error?.Throw();
+            return _result;
+        }
+
+        void IValueTaskSource.GetResult(short token)
+        {
+            GetResult(token);
+        }
+
+        public void Reset()
+        {
+            _completed = false;
```

Comment:
Should this throw if it hasn't been consumed?

Reply:
It could.  This method is never actually called currently :)